### PR TITLE
fix(runtime): persist consumed middleware tool results

### LINKED ISSUE #5670: [bug] Middleware-short-circuited tool results can vanish when a later run step fails
### Before you start

- [x] I searched existing issues and pull requests and did not find an equivalent report or implementation.
- [ ] I have not run the proposed regression in a full local DeerFlow environment yet; the failure below is derived from the current callback flow and should be verified red before implementation.

### Problem summary

A user-visible `ToolMessage` produced by middleware that short-circuits a tool call can still disappear from durable history when the same run later fails.

Current `RunJournal` already handles the successful-run version of this problem: if middleware answers a tool call without invoking the real tool handler, LangChain emits no `on_tool_end`, so `RunJournal.on_chain_end()` reconciles missing current-run `ToolMessage` objects from the root graph's final output.

That fallback only runs on successful root completion.

For middleware that returns a `ToolMessage` and then allows the agent loop to continue (for example a guardrail denial or read-before-write denial), the next model call receives that `ToolMessage` in `on_chat_model_start()`. `RunJournal` currently ignores tool messages at that boundary. If the later model call/run then fails, `on_chain_error()` has no graph outputs to reconcile and root `on_chain_end()` never runs.

The result can therefore be visible during the live run and consumed by the model, but never reach the run-event message feed used by reload/history APIs.

### Current flow

```text
lead model emits tool call
  -> RunJournal.on_llm_end remembers current-run lead tool_call_id

middleware short-circuits the call
  -> returns visible ToolMessage
  -> real handler does not run
  -> no RunJournal.on_tool_end

agent continues
  -> next RunJournal.on_chat_model_start receives the ToolMessage in model history
  -> current code only captures HumanMessage input there

later model/run failure
  -> RunJournal.on_chain_error records run.error
  -> no outputs are available
  -> root on_chain_end is never called
  -> _reconcile_final_tool_messages never sees the ToolMessage

GET /api/threads/{thread_id}/messages/page
  -> reads RunEventStore, not checkpoint state
  -> short-circuited ToolMessage is absent after reload
```

This is distinct from the existing successful-run reconciliation: that path is correct and should remain as the terminal fallback for short-circuits such as `ask_clarification` that end the graph immediately and therefore have no subsequent model call.

### Deterministic regression

A focused callback-level regression can be added to `backend/tests/test_run_journal.py` using the existing helpers:

```python
@pytest.mark.anyio
async def test_model_start_persists_short_circuited_tool_result_before_later_run_error(
    journal_setup,
):
    from langchain_core.messages import ToolMessage

    journal, store = journal_setup

    journal.on_llm_end(
        _make_llm_response(
            "",
            tool_calls=[
                {
                    "id": "call_write",
                    "name": "write_file",
                    "args": {"path": "/tmp/a", "content": "x"},
                }
            ],
        ),
        run_id=uuid4(),
        parent_run_id=None,
        tags=["lead_agent"],
    )

    blocked = ToolMessage(
        id="blocked-write",
        content="Read the file before modifying it.",
        tool_call_id="call_write",
        name="write_file",
        status="error",
    )

    journal.on_chat_model_start(
        {},
        [[blocked]],
        run_id=uuid4(),
        tags=["lead_agent"],
    )

    journal.on_chain_error(RuntimeError("provider failed"), run_id=uuid4())

    await journal.flush()

    messages = await store.list_messages("t1")
    tool_results = [
        event for event in messages
        if event["event_type"] == "llm.tool.result"
    ]
    assert len(tool_results) == 1
    assert tool_results[0]["content"]["id"] == "blocked-write"
```

On current main, `on_chat_model_start()` only persists the first eligible `HumanMessage`, while `on_chain_error()` only writes `run.error`, so there is no code path that can satisfy the `llm.tool.result` assertion above.

Before implementation, please run this regression against current main and treat a non-red result as a stop condition.

### Expected behavior

Once a current-run, lead-agent, user-visible middleware result has become part of a subsequent model request, it should already be durable. A later model/provider failure must not make that previously visible result disappear from history.

The existing safety boundaries should remain intact:

- do not persist hidden `ToolMessage` objects;
- do not attribute retained results from earlier runs to the current run;
- do not move subagent-internal tool results into the parent thread feed;
- do not duplicate results already captured by `on_tool_end`;
- keep root `on_chain_end` reconciliation for terminal short-circuit paths with no next model call.

### Suggested implementation direction

Reuse the existing reconciliation policy rather than adding tool-name-specific handling.

One focused approach:

1. Extract the current `_reconcile_final_tool_messages()` loop into a helper that can reconcile a sequence of messages using the existing `_should_reconcile_tool_message()` guard.
2. In `on_chat_model_start()`, reconcile missing eligible `ToolMessage` objects present in the model-input batches before processing the human-input capture.
3. Keep `on_chain_end()` as the final fallback for short-circuits that terminate the graph immediately.
4. Update the run-event contract/docs if necessary so `llm.tool.result` producers describe both normal `on_tool_end` capture and reconciliation.

### Regression matrix

Please cover at least:

- current-run short-circuited result is durable by the next model start;
- later `on_chain_error` does not erase that history;
- a result already persisted by `on_tool_end` is not duplicated at model start;
- retained old-run tool messages are ignored;
- hidden tool messages are ignored;
- subagent tool-call IDs are ignored;
- existing root-chain-end clarification reconciliation still works unchanged.

### Scope / non-goals

- No persistence architecture change.
- No middleware-specific allowlist.
- No changes to tool authorization semantics.
- No changes to checkpoint compaction behavior.
- No changes to frontend history merging.

### PR DESCRIPTION
**Pull request by KevinAndrewDong**
Fixes #5670

## Why

Middleware can return a user-visible ToolMessage without invoking the real tool handler. LangChain then emits no on_tool_end, and the existing reconciliation only runs on successful root on_chain_end. If a later model call consumes that result and the run fails, the result can be visible live but missing from durable run history.

## What changed

- Reconcile eligible consumed ToolMessage objects at the next lead-agent on_chat_model_start.
- Reconcile only the suffix after the latest current-run AI tool-call boundary in both model-start and root-end paths, so retained old-run results cannot win when a tool_call_id is reused.
- Reuse the existing visibility, current-run ownership, and de-duplication guards.
- Keep root on_chain_end reconciliation for terminal short-circuit paths.
- Document the callback boundary.

## Surface area

- [ ] Frontend UI
- [ ] Backend API
- [x] Agents / LangGraph
- [ ] Sandbox
- [ ] Skills
- [ ] Dependencies
- [x] Default behavior change
- [ ] Docs / tests / CI only

## Screenshots / Recording

Not applicable; no UI surface.

## Bug fix verification

- Regression: `TestFinalToolMessageReconciliation::test_model_start_persists_consumed_short_circuited_tool_result_before_later_run_error`
- The regression was red on `main@656db122` and green on this branch.
- Coverage includes de-duplication, retained old-run messages, AI-present/no-current-run-boundary histories, hidden results, subagent isolation, reused `tool_call_id` collisions in both reconciliation paths, and root-end fallback.

## Validation

- `cd backend && uv run pytest tests/test_run_journal.py -q` — 126 passed
- `cd backend && uv run pytest tests/test_run_event_stream_contract.py tests/blocking_io/test_run_journal_callbacks.py -q` — 40 passed
- Ruff format/check passed
- `git diff --check` passed

## AI assistance

**Tool(s) used:** ChatGPT and Codex

**How you used it:** AI assistance was used for repository analysis, implementation support, and test validation.

- [x] I've read and understand every line of this change and take responsibility for it — it's not unreviewed AI output.

### CODE REVIEW

**Review by willem-bd — COMMENTED** [review]
Reviewed at b2d9c3ad. This matches the approach suggested in #5670 and keeps all four safety boundaries from the issue intact (hidden results, old-run ownership, subagent isolation, on_tool_end dedup, root-end terminal fallback). I verified the suffix invariant holds: a short-circuited result sitting before the latest current-run boundary was necessarily presented to the model call that produced that boundary, so it was already reconciled at that call's on_chat_model_start; the root-end path therefore only needs the terminal suffix. Ran `tests/test_run_journal.py` (125 passed), `tests/test_run_event_stream_contract.py` (39 passed), and `tests/blocking_io/test_run_journal_callbacks.py` (1 passed) at this head, and ruff format/check is clean. Two non-blocking suggestions below.

File: backend/packages/harness/deerflow/runtime/journal.py [hunk 1]
```diff
@@ -695,13 +699,44 @@ def _should_reconcile_tool_message(self, message: ToolMessage) -> bool:
...
+            has_ai_message = True
+            tool_calls = getattr(message, "tool_calls", None) or []
+            if any(isinstance(tool_call_id := self._tool_call_value(tool_call, "id"), str) and tool_call_id in self._current_run_tool_call_names for tool_call in tool_calls):
+                boundary = index
+        if boundary is not None:
+            return batch[boundary + 1 :]
+        # Some callback fixtures provide only the tool result. Preserve that
+        # fallback, but never scan a history containing AI messages without a
+        # current-run boundary.
+        return batch if not has_ai_message else None
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/runtime/journal.py:743]: This skip has no direct regression coverage: every new model-start test either has no AI message in the batch (`test_model_start_ignores_retained_old_run_tool_result` exercises the `not has_ai_message` whole-batch fallback), has a valid current-run boundary, or is a non-lead caller. A lead-agent start shaped like `[[ai_without_current_run_calls, old_tool]]` (AI present, no boundary) would pin this `None` branch. Today its observable outcome coincides with the `_should_reconcile_tool_message` membership guard, so a future edit to either guard could silently change behavior without a test noticing — worth one more case in the matrix given the comment here documents the invariant explicitly.

File: backend/packages/harness/deerflow/runtime/journal.py [hunk 2]
```diff
@@ -695,13 +699,44 @@ def _should_reconcile_tool_message(self, message: ToolMessage) -> bool:
...
+        batch = list(messages)
+        boundary = None
+        has_ai_message = False
+        for index, message in enumerate(batch):
+            is_ai_message = isinstance(message, AIMessage) or getattr(message, "type", None) == "ai"
+            if not is_ai_message:
+                continue
+            has_ai_message = True
+            tool_calls = getattr(message, "tool_calls", None) or []
+            if any(isinstance(tool_call_id := self._tool_call_value(tool_call, "id"), str) and tool_call_id in self._current_run_tool_call_names for tool_call in tool_calls):
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/runtime/journal.py]: Nit (non-blocking): this one-liner packs three concerns — AI-message detection, tool-call id extraction, and current-run membership — into a single walrus-inside-`any()` expression, and the AI-message test on line 721 duplicates the same check `_remember_current_run_tool_calls` does. A tiny shared helper (e.g. `_has_current_run_tool_call(message) -> bool`) would put the boundary rule in one place and make both call sites easier to audit; this helper is subtle enough (it is the fix for #5670) that cheaper reading is real value. Ruff format is happy with the line as-is, so this is purely readability.

File: backend/packages/harness/deerflow/runtime/journal.py (hunk 1 above)
Comment (KevinAndrewDong) [reply backend/packages/harness/deerflow/runtime/journal.py:743]: Added a regression covering the AI-present/no-current-run-boundary case, including a reused current-run tool_call_id so the test would fail if the suffix guard were removed.

File: backend/packages/harness/deerflow/runtime/journal.py (hunk 2 above)
Comment (KevinAndrewDong) [reply backend/packages/harness/deerflow/runtime/journal.py]: Extracted the AI-message/current-run-tool-call checks into small helpers and reused the AI-message helper in _remember_current_run_tool_calls().
